File: nim/robot-name/robot_name.nim
import random, sets

randomize()

type
  Robot* = object
    name*: string

const NumberOfLetters = 2
const NumberOfDigits = 3

const Letters = 'A'..'Z'
const Digits = '0'..'9'

var usedNames = initHashSet[string]()

proc newName: string =
  for x in 0 ..< NumberOfLetters:
    result.add(rand(Letters))
  for x in 0 ..< NumberOfDigits:
    result.add(rand(Digits))

proc randomName: string =
  while true:
    result = newName()

    if result notin usedNames:
      usedNames.incl(result)
      break

proc makeRobot*: Robot =
  result.name = randomName()

proc reset*(robot: var Robot): void =
  robot.name = randomName()

File: r/robot-name/robot-name.R
new_robot <- function() {
  structure(
    list(
      name = random_name()
    ),
    class = "robot"
  )
}

reset_robot <- function(robot) {
  robot$name <- random_name()
  robot
}

random_name <- function() {
  paste0(
    c(sample(LETTERS, 2, replace = TRUE), sample(0:9, 3, replace = TRUE)),
    collapse = ""
  )
}
